File: go/07-estado-contexto/truncacion.go
// Cómo ejecutar: make go FILE=go/07-estado-contexto/truncacion.go
package main

import (
	"encoding/json"
	"fmt"
	"strings"
)

type Message map[string]interface{}

func estimateTokens(messages []Message) int {
	total := 0
	for _, m := range messages {
		b, _ := json.Marshal(m)
		total += len(b)
	}
	return total / 4
}

func validarParidad(messages []Message) []string {
	uses    := map[string]bool{}
	results := map[string]bool{}

	for _, msg := range messages {
		content, ok := msg["content"].([]interface{})
		if !ok {
			continue
		}
		for _, block := range content {
			b, ok := block.(map[string]interface{})
			if !ok {
				continue
			}
			if b["type"] == "tool_use" {
				if id, ok := b["id"].(string); ok {
					uses[id] = true
				}
			} else if b["type"] == "tool_result" {
				if id, ok := b["tool_use_id"].(string); ok {
					results[id] = true
				}
			}
		}
	}

	var orphans []string
	for id := range uses {
		if !results[id] {
			orphans = append(orphans, id)
		}
	}
	for id := range results {
		if !uses[id] {
			orphans = append(orphans, id)
		}
	}
	return orphans
}

func hasToolUse(msg Message) bool {
	content, ok := msg["content"].([]interface{})
	if !ok {
		return false
	}
	for _, block := range content {
		b, ok := block.(map[string]interface{})
		if ok && b["type"] == "tool_use" {
			return true
		}
	}
	return false
}

func truncarFifo(messages []Message, maxTokens int) []Message {
	working := make([]Message, len(messages))
	copy(working, messages)

	for estimateTokens(working) > maxTokens && len(working) > 1 {
		removed := working[0]
		working = working[1:]
		if len(working) > 0 &&
			removed["role"] == "assistant" &&
			hasToolUse(removed) &&
			working[0]["role"] == "user" {
			working = working[1:]
		}
	}
	return working
}

func truncarHeadTail(messages []Message, maxTokens, head, tail int) []Message {
	if len(messages) <= head+tail {
		return messages
	}
	result := make([]Message, 0, head+tail)
	result = append(result, messages[:head]...)
	result = append(result, messages[len(messages)-tail:]...)

	if estimateTokens(result) > maxTokens {
		result = truncarFifo(result, maxTokens)
	}
	return result
}

func limpiarToolResults(messages []Message, minAge int) []Message {
	toolResultCount := 0
	resultMsgs := make([]Message, len(messages))
	copy(resultMsgs, messages)

	for i := len(messages) - 1; i >= 0; i-- {
		msg := messages[i]
		content, ok := msg["content"].([]interface{})
		if !ok {
			resultMsgs[i] = msg
			continue
		}

		newBlocks := make([]interface{}, len(content))
		copy(newBlocks, content)

		for j, block := range content {
			b, ok := block.(map[string]interface{})
			if !ok {
				continue
			}
			if b["type"] == "tool_result" {
				toolResultCount++
				if toolResultCount > minAge {
					newBlock := map[string]interface{}{}
					for k, v := range b {
						newBlock[k] = v
					}
					newBlock["content"] = []interface{}{map[string]interface{}{"type": "text", "text": "[cleared]"}}
					newBlocks[j] = newBlock
				}
			}
		}

		newMsg := Message{}
		for k, v := range msg {
			newMsg[k] = v
		}
		newMsg["content"] = newBlocks
		resultMsgs[i] = newMsg
	}

	return resultMsgs
}

type AlmacenHistorial struct {
	MaxTokens int
	messages  []Message
}

func NewAlmacenHistorial(maxTokens int) *AlmacenHistorial {
	if maxTokens == 0 {
		maxTokens = 110_000
	}
	return &AlmacenHistorial{MaxTokens: maxTokens, messages: []Message{}}
}

func (a *AlmacenHistorial) Add(message Message) {
	a.messages = append(a.messages, message)
}

func (a *AlmacenHistorial) Get() []Message {
	result := make([]Message, len(a.messages))
	copy(result, a.messages)
	return result
}

func (a *AlmacenHistorial) Tokens() int {
	return estimateTokens(a.messages)
}

func (a *AlmacenHistorial) Len() int {
	return len(a.messages)
}

func (a *AlmacenHistorial) ApplyFifo() {
	a.messages = truncarFifo(a.messages, a.MaxTokens)
}

func (a *AlmacenHistorial) ApplyHeadTail(head, tail int) {
	a.messages = truncarHeadTail(a.messages, a.MaxTokens, head, tail)
}

func (a *AlmacenHistorial) ClearToolResults(minAge int) {
	a.messages = limpiarToolResults(a.messages, minAge)
}

func (a *AlmacenHistorial) CheckParity() []string {
	return validarParidad(a.messages)
}

func main() {
	historial := NewAlmacenHistorial(2_000)

	historial.Add(Message{"role": "user", "content": "Analiza el repo."})
	for i := 0; i < 8; i++ {
		id := fmt.Sprintf("tu_%02d", i)
		historial.Add(Message{
			"role": "assistant",
			"content": []interface{}{
				map[string]interface{}{
					"type":  "tool_use",
					"id":    id,
					"name":  "read_file",
					"input": map[string]interface{}{"path": fmt.Sprintf("file_%d.py", i)},
				},
			},
		})
		historial.Add(Message{
			"role": "user",
			"content": []interface{}{
				map[string]interface{}{
					"type":        "tool_result",
					"tool_use_id": id,
					"content":     []interface{}{map[string]interface{}{"type": "text", "text": strings.Repeat("x", 200)}},
				},
			},
		})
	}

	fmt.Printf("Antes: %d msgs, ~%d tokens\n", historial.Len(), historial.Tokens())
	parity := historial.CheckParity()
	if len(parity) == 0 {
		fmt.Println("Paridad: OK")
	} else {
		fmt.Printf("Paridad: %v\n", parity)
	}

	historial.ClearToolResults(4)
	fmt.Printf("\nTras clear_tool_results(min_age=4): ~%d tokens\n", historial.Tokens())
	parity = historial.CheckParity()
	if len(parity) == 0 {
		fmt.Println("Paridad tras limpiar: OK")
	} else {
		fmt.Printf("Paridad tras limpiar: %v\n", parity)
	}

	historial.ApplyHeadTail(1, 4)
	fmt.Printf("\nTras head_tail(1,4): %d msgs, ~%d tokens\n", historial.Len(), historial.Tokens())
	parity = historial.CheckParity()
	if len(parity) == 0 {
		fmt.Println("Paridad final: OK")
	} else {
		fmt.Printf("Paridad final: %v\n", parity)
	}
}

File: typescript/07-estado-contexto/truncacion.ts
// Cómo ejecutar: make ts SCRIPT=typescript/07-estado-contexto/truncacion.ts
type Message = Record<string, unknown>;

function estimateTokens(messages: Message[]): number {
  return messages.reduce((sum, m) => sum + JSON.stringify(m).length, 0) / 4;
}

function validarParidad(messages: Message[]): string[] {
  const uses    = new Set<string>();
  const results = new Set<string>();

  for (const msg of messages) {
    const content = msg["content"];
    if (!Array.isArray(content)) continue;
    for (const block of content) {
      if (typeof block !== "object" || block === null) continue;
      const b = block as Record<string, unknown>;
      if (b["type"] === "tool_use" && typeof b["id"] === "string") {
        uses.add(b["id"]);
      } else if (b["type"] === "tool_result" && typeof b["tool_use_id"] === "string") {
        results.add(b["tool_use_id"]);
      }
    }
  }

  const orphans: string[] = [];
  for (const id of uses)    if (!results.has(id)) orphans.push(id);
  for (const id of results) if (!uses.has(id))    orphans.push(id);
  return orphans;
}

function hasToolUse(msg: Message): boolean {
  const content = msg["content"];
  if (!Array.isArray(content)) return false;
  return content.some(
    (b) => typeof b === "object" && b !== null && (b as Record<string, unknown>)["type"] === "tool_use"
  );
}

function truncarFifo(messages: Message[], maxTokens: number): Message[] {
  const working = [...messages];
  while (estimateTokens(working) > maxTokens && working.length > 1) {
    const removed = working.shift()!;
    if (
      working.length > 0 &&
      removed["role"] === "assistant" &&
      hasToolUse(removed) &&
      working[0]["role"] === "user"
    ) {
      working.shift();
    }
  }
  return working;
}

function truncarHeadTail(
  messages: Message[],
  maxTokens: number,
  head = 2,
  tail = 6
): Message[] {
  if (messages.length <= head + tail) return messages;

  let result = [...messages.slice(0, head), ...messages.slice(messages.length - tail)];
  if (estimateTokens(result) > maxTokens) {
    result = truncarFifo(result, maxTokens);
  }
  return result;
}

function limpiarToolResults(messages: Message[], minAge = 4): Message[] {
  let toolResultCount = 0;
  const resultMsgs: Message[] = [];

  for (let i = messages.length - 1; i >= 0; i--) {
    const msg     = messages[i];
    const content = msg["content"];

    if (!Array.isArray(content)) {
      resultMsgs.unshift(msg);
      continue;
    }

    const newBlocks = content.map((block) => {
      if (
        typeof block === "object" &&
        block !== null &&
        (block as Record<string, unknown>)["type"] === "tool_result"
      ) {
        toolResultCount += 1;
        if (toolResultCount > minAge) {
          return { ...(block as object), content: [{ type: "text", text: "[cleared]" }] };
        }
      }
      return block;
    });

    resultMsgs.unshift({ ...msg, content: newBlocks });
  }

  return resultMsgs;
}

class AlmacenHistorial {
  maxTokens: number;
  private _messages: Message[];

  constructor(maxTokens = 110_000) {
    this.maxTokens = maxTokens;
    this._messages = [];
  }

  add(message: Message): void {
    this._messages.push(message);
  }

  get(): Message[] {
    return [...this._messages];
  }

  get tokens(): number {
    return estimateTokens(this._messages);
  }

  get length(): number {
    return this._messages.length;
  }

  applyFifo(): void {
    this._messages = truncarFifo(this._messages, this.maxTokens);
  }

  applyHeadTail(head = 2, tail = 6): void {
    this._messages = truncarHeadTail(this._messages, this.maxTokens, head, tail);
  }

  clearToolResults(minAge = 4): void {
    this._messages = limpiarToolResults(this._messages, minAge);
  }

  checkParity(): string[] {
    return validarParidad(this._messages);
  }
}

const historial = new AlmacenHistorial(2_000);

historial.add({ role: "user", content: "Analiza el repo." });
for (let i = 0; i < 8; i++) {
  historial.add({
    role:    "assistant",
    content: [{ type: "tool_use", id: `tu_${String(i).padStart(2, "0")}`, name: "read_file", input: { path: `file_${i}.py` } }],
  });
  historial.add({
    role:    "user",
    content: [{ type: "tool_result", tool_use_id: `tu_${String(i).padStart(2, "0")}`, content: [{ type: "text", text: "x".repeat(200) }] }],
  });
}

console.log(`Antes: ${historial.length} msgs, ~${historial.tokens} tokens`);
console.log(`Paridad: ${historial.checkParity().length === 0 ? "OK" : historial.checkParity()}`);

historial.clearToolResults(4);
console.log(`\nTras clearToolResults(minAge=4): ~${historial.tokens} tokens`);
console.log(`Paridad tras limpiar: ${historial.checkParity().length === 0 ? "OK" : historial.checkParity()}`);

historial.applyHeadTail(1, 4);
console.log(`\nTras head_tail(1,4): ${historial.length} msgs, ~${historial.tokens} tokens`);
console.log(`Paridad final: ${historial.checkParity().length === 0 ? "OK" : historial.checkParity()}`);
